### DISCUSSION TITLE
20221121 developer call notes

### DISCUSSION THREAD

**Discussion by hammer** [sgkit-dev/sgkit#954] [Meeting Notes]
### Pre-notes

* [Notes](https://github.com/pystatgen/sgkit/discussions/950) from previous meeting

#### Issues

* [sgkit#692](https://github.com/pystatgen/sgkit/issues/692) Implement gene set association analysis (gene-ε)
* [sgkit-publication#2](https://github.com/pystatgen/sgkit-publication/issues/2) Agree on outline

### Notes

* This meeting will focus on nailing down an outline for the paper!

#### Attendees

1. Jeff Hammerbacher
2. Tom White
3. Ben Jeffery
4. Eric Czech
5. Lorin Crawford
6. Wei Cheng
7. Jerome Kelleher (late)

#### Discussion

* Wei now at Meta, formerly with Lorin working on gene-ε
* [sgkit#692](https://github.com/pystatgen/sgkit/issues/692)
    * TW worked on feasibility and requirements for parallelization
    * TW feels we can get a long way with a big machine
    * EC: gene-ε for binary outcomes?
    * LC: method works mostly the same, additional corrections weren’t critical
    * WC: ran some simulations to validate
    * EC: motivation for Gaussian mixture modeling for inferring clusters?
    * LC: Nilanjan Chatterjee’s JHU group found distributions of SNP effect sizes are complicated (JH: [Estimation of complex effect-size distributions using summary-level statistics from genome-wide association studies across 32 complex traits](https://pubmed.ncbi.nlm.nih.gov/30104760/)). Need to make SNP-level effect size distributions look like generative model. Near-null SNPs are the ones you want to shrink to zero. “Put all SNPs back into their appropriate classes”. Gene-ε tries to learn a > 0 threshold to pull additional SNPs back to zero.
    * LC: interpretation of clusters would be useful to understand but haven’t solved it yet.
    * EC: Okay we can just hide the clusters from users.
    * EC: What’s missing from sgkit? LD matrices are formalized as a data structure with documentation that could be populated from external sources.
        * JH: [previous discussion](https://github.com/pystatgen/sgkit/issues/279#issuecomment-840664157) of external LD matrices in sgkit
        * WC: use biglasso’s support for loading LD matrices
    * TW: scikit-learn probably sufficient for GLM step.
    * EC: what shrinkage method is best?
    * LC: we used Elastic Net. (JH: [ElasticNet](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.ElasticNet.html) in sklearn)
    * EC: L2 easier to scale; will it be okay?
    * WC: probably not, need L1
    * TODO: find scalable L1
    * LC: most methods run per-chromosome for scalability
    * LC: [RSS](https://github.com/stephenslab/rss) was the closest competitor to gene-ε
    * EC: limitations on how you could re-group for pathways?
    * LC: Nope. Using gene annotations now, but could pass other annotations.
    * LC: Have minimum number of SNPs in a set
* [sgkit-publication#2](https://github.com/pystatgen/sgkit-publication/issues/2)
    * JK: stuff everything into Results
    * JK: structure around 5 - 6 case studies
    * JK: worried functionality section will be dull; expect readers to jump to “Does it do what I want to do? Does it do it fast?”	
    * JK: don’t want to miss pairwise stuff on GPU that Amit Kumar did
    * BJ: data set in memory and on disk improves interoperability
    * JH: our “duck typing” approach on top of xarray
    * JH: what have we worked hard on?
        * TW: MalariaGEN use case run on multi-core single machine
        * TW: windowing work
        * EC: NVIDIA work on REGENIE → computing LD on giant cohorts with scalable ExWAS
    * TW: would be nice for use cases to use heterogeneous back ends, e.g. multi-core single machine, cluster, GPUs
    * JK: how to highlight the openness of the data structure?
        * JH: 3 standards, e.g. on top of Xarray, VCF in Zarr, file formats
        * JK: tsinfer writing additional columns to Zarr for tskit
    * Use case enumeration
        * BJ: Tskit/tsinfer add additional columns to VCF and write as Zarr
        * TW: MalariaGEN pop gen (including windowing)
        * T M: Tim Millar’s poly-ploidy work w/ pedigrees
        * JH/EC: Canine GWAS
        * AK/JK: Amit’s GPU work doing pairwise distances to build a tree
        * JH: NVIDIA’s GPU work to speed REGENIE
        * TW: gene-ε
    * TW: benchmarks against Hail and PLINK?
        * JH and JK: don’t focus on it
    * TW: EC hit lots of GWAS scalability issues on synthetic data
        * JH: can we use Oxford work on UKB data in perf section?
        * TW: compression work with Ben Ellsworth
    * JH: Should we do use cases as “Examples” in our docs first or second?
        * Pros: notebook as rough draft, notebook as reference from paper
        * Cons: more time, examples a little different
    * JK: What are our display items?
        * JK: Most important figure should be Figure 1
        * JH: EC’s data model
        * JK: Visualize PyData ecosystem somehow?
            * JH: NumFOCUS PyData ecosystem figure?
        * JH: Visualize API structure
        * BJ: scaling performance graphic? In core, out of core, distributed.
            * JH: what method would we use?
            * TW: summary statistics
            * JH: pull in Liangde’s work
        * JH figure order
            * 1: PyData ecosystem
            * 2: Data model
            * 3: API
            * 4: Scale
    * JH: Mention TW’s work on scale out? TW: let’s do Dask first.
    * TW: [New Working Group for Distributed Array Computing](https://discourse.pangeo.io/t/new-working-group-for-distributed-array-computing/2734/1) has lots of alternatives!
* [xarray#7292](https://github.com/pydata/xarray/issues/7292) Xarray doesn’t handle Zarr’s fill_value correctly
    * BJ encountered 
    * JH asked for sgkit issue w/ upstream label
